### PR TITLE
Send fault message on security exception

### PR DESCRIPTION
**Pull request by jonlouie**
### Description
Closes: #676

This PR sets the security exception if there is a failure during security validation. Previously, this exception was set to null and never updated, resulting in the fault message not being sent.

---

### Testing
* Add negative test to ensure that the correct fault exception is thrown when user name validation fails

### CODE REVIEW

**Review by birojnayak — APPROVED** [review]
looks ok

**Comment by mconnew** [comment]
I think you just helped find a bug which has been causing a flakey test. There's another small change that's needed. `SendFaultIfRequired` needs to be made async and the calls to `innerContext.ReplyAsync(faultMessage);` and `innerContext.CloseAsync(timeoutHelper.GetCancellationToken());` need to be awaited. We've had a flakey test which sometimes works and sometimes doesn't in the failure code path, and that's because `CloseAsync` has been aborting the connection before the fault message had finsihed sending. There's a second similar case on line 642 (where it's called) and line 648 where there's a similarly named method which isn't awaiting on the SendAsync (and should also be calling CloseAsync after sending).

File: src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs [hunk 1]
```diff
@@ -640,13 +639,13 @@ private async ValueTask<Message> ProcessInnerItemAsync(Message innerItem, TimeSp
...
             {
-                SendFaultIfRequired(securityException, unverifiedMessage, timeoutHelper.RemainingTime());
+                await SendFaultIfRequired(securityException, unverifiedMessage, timeoutHelper.RemainingTime());
                 return null;
             }
             return innerItem;
         }
 
-        private void SendFaultIfRequired(Exception e, Message unverifiedMessage, TimeSpan timeout)
+        private async Task SendFaultIfRequired(Exception e, Message unverifiedMessage, TimeSpan timeout)
```
Comment (mconnew) [inline src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs:648]: Can you change the method names that are now async to have the suffix Async?

Comment (jonlouie) [reply src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs:648]: Ah, good catch. WIll fix

File: src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs [hunk 2]
```diff
@@ -663,7 +662,9 @@ private void SendFaultIfRequired(Exception e, Message unverifiedMessage, TimeSpa
                 {
                     if (unverifiedMessage.Headers.MessageId != null)
                         faultMessage.InitializeReply(unverifiedMessage);
-                    ((IDuplexChannel)InnerDuplexChannel).SendAsync(faultMessage);
+                    TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+                    await ((IDuplexChannel)InnerDuplexChannel).SendAsync(faultMessage);
+                    await ((IDuplexChannel)InnerDuplexChannel).CloseAsync(timeoutHelper.GetCancellationToken());
                 }
```
Comment (mconnew) [inline src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs:668]: There are some test failures with the error message "Expected dispatcher state to be Opened or Faulted, instead it's Closed", so I think I was mistaken in saying that CloseAsync needs to be added here. The difference between this method and the earlier one is this is for duplex channels so it's not appropriate to close as the channel itself didn't fault, it was just the message was bad.

Comment (jonlouie) [reply src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs:667]: That makes sense. Fixed and squashed the commits in the latest push

**Pull request merged**